Hierarchy view styling

grzesiek2010 opened the pull request:
Closes #5837 

#### Why is this the best possible solution? Were any other approaches considered?
As discussed in the issue, the hierarchy view and margins used there should be consistent with what we have in form filling. To achieve that I used the same values defined in the `dimen` file.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The pr is focused on UI improvements that shouldn't be risky for the functionality, therefore I think we can focus on the UI and not perform a lot of regression tests here. It contains some refactoring so it would be good to perform some of them though.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 committed:
Moved the code used to build form hierarchy to one package

grzesiek2010 committed:
Converted HierarchyElement to kotlin

grzesiek2010 committed:
Converted ViewOnlyFormHierarchyActivity to kotlin

grzesiek2010 committed:
Converted HierarchyListAdapter to kotlin

grzesiek2010 committed:
Reworked the hierarchy_element layout using ConstraintLayout

grzesiek2010 committed:
Indent the answer

grzesiek2010 committed:
Use consistent font style in both form entry and form hierarchy

grzesiek2010 committed:
Reworked the hierarchy layout and added the group icon

grzesiek2010 committed:
Fixed group title margin in form entry

grzesiek2010 committed:
Fixed the width of elements in the form hierarchy

grzesiek2010 committed:
Fixed imports

alyblenkin commented:
@grzesiek2010 - thanks for fixing the build so I could review it! The hierarchy screen and form entry are feeling much more unified. 

Notes:
- **Padding:** looks much better, it doesn't feel as squished anymore. 

- **Spacing between questions:** now that I'm seeing it built out, I think we need to increase the spacing between questions just a little bit. I think this is making it harder to read when you have a longer form. 

- **Folder icon:** I'm quite pleased with the icon! I think it helps with navigation and establishing a hierarchy of information and my eye is drawn to the top first. I think we could add a bit more padding between the group name and question to establish more of a separation between the two.

- **Font size and weight:** I think the question looks good, but the answer seems a bit too small. Can we increase it slightly? I'm also thinking we may want to change the text colour of the answer to blue to make it more obvious - [figma](https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3936%3A74573&mode=design&t=xcR7SjhvHFtKrwjy-1). This could look terrible, but I'm curious to see if that helps.

- **Indenting answer:** Indenting certainly helps with scanning the information when I compare it to what it looked like before. However, it's not working consistently across all question types. The answers for the required questions seem to appear aligned to the left, rather than indented (see below). Side note: Now that the answer is indented, it's very noticeable that when the required question falls onto two lines it doesn't get indented (so everything comes after the asterisk).

- **Table of contents / summary view:** I don't think this should be indented because it makes it harder to read - I'm thinking that styling accidentally was applied there too. I also still think the [label and name should be flipped](https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3684%3A73031&mode=design&t=xcR7SjhvHFtKrwjy-1), but we didn't prioritize it as part of this story. Maybe this is something we come back to. 

<img width="667" alt="Screenshot 2024-01-03 at 9 43 28 AM" src="https://github.com/getodk/collect/assets/7871340/7382b3fd-33c7-43f2-bfef-dbc28a8b830e">

grzesiek2010 commented:
> Indenting answer: Indenting certainly helps with scanning the information when I compare it to what it looked like before. However, it's not working consistently across all question types. The answers for the required questions seem to appear aligned to the left, rather than indented (see below). Side note: Now that the answer is indented, it's very noticeable that when the required question falls onto two lines it doesn't get indented (so everything comes after the asterisk).

The asterisk is a part of a question text. We don't have a separate view to display the asterisk we just display it in one TextView. Since it's one TextView if it consists of multiple lines they are not aligned vertically with the first line but with the asterisk that in fact is the beginning of the first line not a separate element as I said above. Do you think it would be better to have it in a separate element? Some time ago I wanted to refactor the way we display asterisks in form entry (they are displayed in the same way by adding them to question labels) but it turned out to be too complex. If you want I can check if it is possible to rework it in the hierarchy view at least.

> Table of contents / summary view: I don't think this should be indented because it makes it harder to read 

Do you mean that the Group text is indented too or maybe something else as well? I agree it shouldn't and maybe a good solution would be to display it above the question as you said https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3684-73031&mode=design&t=z5x1vMKVcsaifguG-0 to make it clear it's not the same as answers in other list elements

> Spacing between questions: now that I'm seeing it built out, I think we need to increase the spacing between questions just a little bit. I think this is making it harder to read when you have a longer form.

Something like this would be better?

| Before  | After |
| ------------- | ------------- |
| ![Screenshot_1704383830](https://github.com/getodk/collect/assets/3276264/38e624b8-4c35-49a9-81ed-c6ea7e7485f8)  | ![Screenshot_1704384808](https://github.com/getodk/collect/assets/3276264/918c2eed-59f2-4170-ad5e-b6de1ec62ffd) |

alyblenkin commented:
> The asterisk is a part of a question text. We don't have a separate view to display the asterisk we just display it in one TextView. Since it's one TextView if it consists of multiple lines they are not aligned vertically with the first line but with the asterisk that in fact is the beginning of the first line not a separate element as I said above. Do you think it would be better to have it in a separate element? Some time ago I wanted to refactor the way we display asterisks in form entry (they are displayed in the same way by adding them to question labels) but it turned out to be too complex. If you want I can check if it is possible to rework it in the hierarchy view at least.

That's helpful context. I think we can address this as a separate issue. I think it wasn't bad before, but more noticeable with the answer indented. I'll write up a story, and we can determine where it fits in the priority later :) 

> Do you mean that the Group text is indented too or maybe something else as well? I agree it shouldn't and maybe a good solution would be to display it above the question as you said https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3684-73031&mode=design&t=z5x1vMKVcsaifguG-0 to make it clear it's not the same as answers in other list elements

I haven't noticed it anywhere else, so just the group text for now. 

> Something like this would be better?

I think that's looking a lot better. The old spacing isn't so bad in this screenshot because the text is really short. The spacing becomes difficult when you have lengthy questions, making it harder to distinguish the separation between things. Do you have an example of when you drill down into the questions with the new spacing?

grzesiek2010 committed:
Increased the vertical padding of hierarchy elements

grzesiek2010 commented:
> I think that's looking a lot better. The old spacing isn't so bad in this screenshot because the text is really short. The spacing becomes difficult when you have lengthy questions, making it harder to distinguish the separation between things. Do you have an example of when you drill down into the questions with the new spacing?

I've just pushed that change so you can download the apk and play with it on your own. Please notice that now the top margin in the form entry will be smaller than the one in the hierarchy view and when you navigate from the form entry to the hierarchy view the question title goes down a little bit. I think we wanted to have it consistent so adjusting the margin in the form entry might be needed too?
![ezgif com-video-to-gif-converter](https://github.com/getodk/collect/assets/3276264/d1a32e1b-5aba-4be7-8fef-294e941b572a)

alyblenkin commented:
> I've just pushed that change so you can download the apk and play with it on your own. 

Thank you! The spacing works a lot better for longer questions as well. Is it possible to reduce the vertical padding just a little bit, so it isn't as drastic when you have shorter text? 

> Please notice that now the top margin in the form entry will be smaller than the one in the hierarchy view and when you navigate from the form entry to the hierarchy view the question title goes down a little bit. I think we wanted to have it consistent so adjusting the margin in the form entry might be needed too? 

I agree, we need to make the top margin consistent so it doesn't jump around.

grzesiek2010 commented:
> Is it possible to reduce the vertical padding just a little bit, so it isn't as drastic when you have shorter text?

on master it was 8dp now it's 16dp so if you want something between we can try 12dp?

alyblenkin commented:
> on master it was 8dp now it's 16dp so if you want something between we can try 12dp?

Yes, something between 10dp or 12dp will work well. Let's try 12dp first. Let me know if it's easier to hop on a call to tweak it either today or Monday.

grzesiek2010 committed:
Reworked hierarchy item views

grzesiek2010 committed:
Increase the font size of answers

grzesiek2010 committed:
Add vertical padding to the list of widgets to match the hierarchy view padding

grzesiek2010 commented:
> Yes, something between 10dp or 12dp will work well. Let's try 12dp first. Let me know if it's easier to hop on a call to tweak it either today or Monday.

I've used 12dp

and also updated the text appearance used for displaying answers:

| Before  | After |
| ------------- | ------------- |
| ![Screenshot_1704718509](https://github.com/getodk/collect/assets/3276264/c87ffafd-91b3-4db8-8ff8-aab8a697aa7e) | ![Screenshot_1704718910](https://github.com/getodk/collect/assets/3276264/0ff22661-bc9b-430e-af62-44f1628c246f) |

the last thing is that I've reworked the way list items for groups are displayed (moving the group label above the title) according to https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&mode=design&t=kITBl6XtaOjSEUri-0

Please review.

> Folder icon: I'm quite pleased with the icon! I think it helps with navigation and establishing a hierarchy of information and my eye is drawn to the top first. I think we could add a bit more padding between the group name and question to establish more of a separation between the two.

Do you still want to add more spacing?

alyblenkin commented:
> I've used 12dp
> and also updated the text appearance used for displaying answers:

Thanks for making those changes! For the spacing, 12dp works well. It's a small change, but it looks much better than the older version, and we are not impacting the experience too much for smaller devices. 

I like the blue text because it makes the answer stand out. I did question if anyone would mistake it for a link, but I doubt that would happen because it's not underlined. Let's quickly review these design updates on the team call tomorrow before we make the final decision.

> the last thing is that I've reworked the way list items for groups are displayed (moving the group label above the title) according to https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&mode=design&t=kITBl6XtaOjSEUri-0

Ah okay, I noted in my[ first comment](https://github.com/getodk/collect/pull/5885#issuecomment-1875742917) that this is something we didn't want to prioritize right now, which is why I didn't include it in the 'dev ready' page that you linked, but I do think it works a lot better! We can discuss this tomorrow as well. 

> Folder icon: I'm quite pleased with the icon! I think it helps with navigation and establishing a hierarchy of information and my eye is drawn to the top first. I think we could add a bit more padding between the group name and question to establish more of a separation between the two.

> Do you still want to add more spacing?

I think this might be okay now that we've spaced out the rest.

alyblenkin commented:
@grzesiek2010 - I also just noticed that the form entry text field is really pushed down now - I don't think that was happening last week. 

![Screenshot_20240108-181654_ODK Collect](https://github.com/getodk/collect/assets/7871340/bf751094-66dc-4e2d-933b-333c7ebd2490)

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formhierarchy/HierarchyListItemView.kt:
```diff
@@ -2,11 +2,11 @@ package org.odk.collect.android.formhierarchy
 import org.odk.collect.android.utilities.HtmlUtils
 
-class HierarchyListItemView(context: Context) : FrameLayout(context) {
+class HierarchyListItemView(context: Context) : ConstraintLayout(context) {
```
I think we still just have `ConstraintLayout` here as `HierarchyListItemView` is just an empty parent and then the layout (a `ConstraintLayout`) is inflated inside it on line 11.

grzesiek2010 replied:
Right, fixed.

seadowg commented on collect_app/src/main/res/values/typography.xml:
```diff
@@ -32,6 +32,10 @@
         <item name="android:textColor">@color/color_on_surface_medium_emphasis</item>
     </style>
 
+    <style name="TextAppearance.QuestionTitle" parent="TextAppearance.Material3.BodyLarge">
```
How about using `?textAppearanceTitleLarge` instead of adding a new typography token? If that's too big, we could customise it so that it's 20sp.

grzesiek2010 replied:
If 22sp is ok then yeah we can use `?textAppearanceTitleLarge` if not then customizing it would basically mean the same as I have done here right so why customizing `TitleLarge` would be better than customizing `BodyLarge`?

Below there is a table the first row contains the version with 22sp and the second one 20sp what do you think @seadowg @alyblenkin?
 
|  |  |
| ------------- | ------------- |
| ![Screenshot_1704833048](https://github.com/getodk/collect/assets/3276264/c7f1d6e7-30ff-4e6b-bf83-2f4ee8ac436b) | ![Screenshot_1704833050](https://github.com/getodk/collect/assets/3276264/56914953-0be9-4c56-be92-75622bf13463) |
| ![Screenshot_1704833096](https://github.com/getodk/collect/assets/3276264/6731ee7f-09c7-4a86-b057-d9df6b7a4fc9) | ![Screenshot_1704833100](https://github.com/getodk/collect/assets/3276264/a9bd61fa-2dbb-462c-be8a-afe9544a1360) |

alyblenkin replied:
I think 20sp works better.

seadowg replied:
> If 22sp is ok then yeah we can use ?textAppearanceTitleLarge if not then customizing it would basically mean the same as I have done here right so why customizing TitleLarge would be better than customizing BodyLarge?

Yeah sorry what I said was pretty unclear. I mean instead of introducing a new style (`QuestionTitle`), we just modify `?textAppearanceTitleLarge` for our theme:

**theme.xml**
```
<item name="textAppearanceTitleLarge">@style/TextAppearance.Collect.TitleLarge</item>
```

**typography.xml**
```
<style name="TextAppearance.Collect.TitleLarge" parent="TextAppearance.Material3.TitleLarge">
    <item name="android:textSize">20sp</item>
</style>
```

grzesiek2010 replied:
Doing things like that with colors for example is ok like we do with:
```
    <style name="TextAppearance.Collect.Headline4" parent="TextAppearance.MaterialComponents.Headline4">
        <item name="android:textColor">@color/color_on_surface_high_emphasis</item>
    </style>
```

but customizing font size and still using the original style name seems misleading to me. There are actually two reasons why I'm not sure about that:
- `textAppearanceTitleLarge` is already used in at least one place (`crash_layout`). Doing what you described would affect that screen. 
- when I look for a good style for a new text I add, then I check https://m3.material.io/styles/typography/type-scale-tokens. In such a case when you look at `TitleLarge` and its 22sp font size you may think that this is what you need. Then you apply this style and surprise something is wrong and you have to analyze what exactly. 

Do you think such font size changing for already existing styles is ok?

seadowg replied:
> Do you think such font size changing for already existing styles is ok?

Good question! Yeah I do. Doing so is referenced in both the [Material 3 docs](https://m3.material.io/styles/typography/type-scale-tokens) and the [Material Components docs](https://github.com/material-components/material-components-android/blob/master/docs/theming/Typography.md#customization). My understanding is that from Material 2 onwards, the philosophy is that instead of adding your own components and styles on top of the Material styles, you're able to "theme" Material so that the standard Material components fit into our "brand". If we change the `?textAppearanceTitleLarge` it will also affect any Material components that might use that text appearance, but that's what we want.

> textAppearanceTitleLarge is already used in at least one place (crash_layout). Doing what you described would affect that screen.

It might be that we want to use a different token there (like a Headline one), or maybe 20sp is also good for that screen!

> when I look for a good style for a new text I add, then I check https://m3.material.io/styles/typography/type-scale-tokens. In such a case when you look at TitleLarge and its 22sp font size you may think that this is what you need. Then you apply this style and surprise something is wrong and you have to analyze what exactly.

That's a good point. When we customise these, we do lose some level of documentation. I think that's ok generally though as we should more the follow the ["roles"](https://m3.material.io/styles/typography/applying-type) when choosing a text appearance for something. I chose "Title Large" here as it seems to me to be the best description of the text. I could imagine that using (and maybe customising) a "Body" token would make more sense though given that question text can be of any length.

grzesiek2010 replied:
Ok but if I have a choice between customizing `BodyLarge` and `TitleLarge` I think it's better to use the latter. There are two reasons:
- `TitleLarge` is used only in one place but `BodyLarge` in many
- `TitleLarge's` original font size is 22sp so the change is not that big as it would be in the case of `BodyLarge` (original front size 16sp)

seadowg replied:
Yeah that makes sense to me. It does feel more like a "title" than "body" text as well.

alyblenkin commented:
Hey @grzesiek2010, so based on everyone's feedback we want to:

- remove the blue text for the answers, but keep the indenting  
- keep the group label above, as you have it now
- keep the spacing at 12pd
- we need to make sure the styling for the answer in the hierarchy and form entry answer is the same

lognaturel commented:
> we need to make sure the styling for the answer in the hierarchy and form entry answer is the same

I was talking about the styling for the question. The answer text is currently bigger when filling out the form than in the hierarchy screen and I think that’s ok.

alyblenkin commented:
> I was talking about the styling for the question. The answer text is currently bigger when filling out the form than in the hierarchy screen and I think that’s ok.

This makes more sense and maybe that's why I was confused because I think the question styling is the same right now, but maybe I'm wrong!

grzesiek2010 committed:
Get rid of the blue color in answers

grzesiek2010 committed:
HierarchyListItemView should extend FrameLayout

grzesiek2010 commented:
>  I also just noticed that the form entry text field is really pushed down now - I don't think that was happening last week.

Fixed.

> remove the blue text for the answers, but keep the indenting

Fixed.

grzesiek2010 committed:
Display proper group icons

alyblenkin commented:
> I also just noticed that the form entry text field is really pushed down now - I don't think that was happening last week.
> Fixed.

> remove the blue text for the answers, but keep the indenting
> Fixed.

Nice! It looks like the "Go To Start" and "Go To End" buttons are now touching. I'm not sure if that was happening before?

grzesiek2010 commented:
> It looks like the "Go To Start" and "Go To End" buttons are now touching. I'm not sure if that was happening before?

Ah yeah my mistake, it should be fixed now.

grzesiek2010 committed:
Use textAppearanceBodyLarge instead of a custom QuestionTitle

alyblenkin reviewed (approved):
Looks great to me!

grzesiek2010 committed:
Moved handling question answers for the hierarchy view to the hierarchy package

grzesiek2010 committed:
Converted QuestionAnswerProcessor to kotlin

grzesiek2010 committed:
Added tests

seadowg commented on collect_app/src/main/java/org/odk/collect/android/formhierarchy/HierarchyItem.kt:
```diff
@@ -0,0 +1,20 @@
+    val primaryText: CharSequence,
+    val secondaryText: String?
+) {
+    constructor(
```
Instead of a second constructor, you should be able to just declare `secondaryText` as optional and generate overloads for Java (if needed):

```kotlin
data class HierarchyItem @JvmOverloads constructor(
   val formIndex: FormIndex,
   val hierarchyItemType: HierarchyItemType,
   val primaryText: CharSequence,
   val secondaryText: String? = null
)
```

grzesiek2010 replied:
Right, fixed.

seadowg commented on collect_app/src/main/res/layout/hierarchy_group_item.xml:
```diff
@@ -0,0 +1,38 @@
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:paddingVertical="12dp"
```
Should we pull out a shared dimension for the `12dp` value? It looks like we're using it in quite a few places? 

I'm also wondering how this ends up being so consistent with the actual question view? Do we end up with other margins adding up to  12dp? If so, maybe we can clean that up so we just end up with one view declaring the spacing.

grzesiek2010 replied:
> Should we pull out a shared dimension for the 12dp value? It looks like we're using it in quite a few places?

Yeah, I was thinking about it as well... I'm not sure about naming though, the value is between `margin_small` (8dp) and `margin_standard` (16dp) so maybe `margin_medium`?

> I'm also wondering how this ends up being so consistent with the actual question view? Do we end up with other margins adding up to 12dp? If so, maybe we can clean that up so we just end up with one view declaring the spacing.

It's 12 because in odk_view we have the group name with `margin_top` = 8dp and then below question widgets with `margin_top` = 4dp, 8+4=12 so that's 12 is the value we needed to use in the hierarchy to display the question title in the same position. Of course it would be better to use the standard margin 16dp but it's too much and and doesn't look good.

seadowg replied:
> Yeah, I was thinking about it as well... I'm not sure about naming though, the value is between margin_small (8dp) and margin_standard (16dp) so maybe margin_medium?

Let's make `12dp` our `margin_small` and `_small` to `_extra_small` and `_extra_small` to `_extra_extra_small`. That will mirror the sizes we have for `_large`.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/res/layout/hierarchy_group_item.xml:
```diff
@@ -0,0 +1,38 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/margin_small"
+        android:textAppearance="@style/TextAppearance.AppCompat.Small"
```
This should use a Material 3 text appearance attribute rather than an explicit style.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/formhierarchy/QuestionAnswerProcessorTest.kt:
```diff
@@ -0,0 +1,26 @@
+import org.mockito.kotlin.whenever
+import org.odk.collect.android.utilities.Appearances
+
+// TODO: Add tests for other question/data types
```
Probably best to avoid adding TODOs as we don't tend to keep track of them anyway. I'm happy either way, but if you think it needs tracked, an issue would be better I think.

grzesiek2010 replied:
Right, it would be good to have an issue for that https://github.com/getodk/collect/issues/5920

seadowg commented on collect_app/src/main/res/layout/hierarchy_repeatable_group_item.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Is there a reason we couldn't just use `hierarchy_group_item` for both of these and change the icon? Maybe I'm missing a detail here.

grzesiek2010 replied:
So far we have had just one hierarchy item that we would adjust by not only setting the primary/secondary text but also by displaying icons. With the newest changes, it would be more difficult so I decided to create separate layouts for every item type. You are right that it would be possible to have one for those two elements (groups) and only set the icon programmatically but it would be a mix of those two approaches. Before it was done programmatically, now it's done using separate layouts and both those approaches are ok. Mixing them doesn't sound that good though so I like more how it's done here.

grzesiek2010 committed:
Fixed conflicts

grzesiek2010 committed:
Use Material 3 text appearance for group labels

grzesiek2010 committed:
Removed redundant constructor

grzesiek2010 committed:
Fixed imports in tests

seadowg reviewed (changes requested):
https://github.com/getodk/collect/pull/5885#discussion_r1455115147

grzesiek2010 committed:
Rename margin_extra_small to margin_extra_extra_small

grzesiek2010 committed:
Rename margin_small to margin_extra_small

grzesiek2010 committed:
Introduced a new margin_small

dbemke commented:
What's the default settings when there are many question in a group and I go to the hierarchy when I'm e.g on a 6th question in the group. What should be at the top of the hierarchy view after going to it from 6th question of a group?
![hierarchy](https://github.com/getodk/collect/assets/56479916/8cb0773b-9adc-455d-b39b-5cba9d6f8c30)

grzesiek2010 commented:
If there are multiple questions in one group when you open the hierarchy view the list should be scrolled to display the question you were on at the top of the list if it's possible so:
- if you open the hierarchy from the first question nothing should change
- if you open the list from the second question, the list should be scrolled so that it's at the top (the same with all the other questions)
- if you are on one of the last questions we just scroll the list to the bottom.

It has been like that, it's not a new behavior.

grzesiek2010 commented:
The only problem I've just noticed is that during scrolling such a long list of question there is no margin between the group name and the list:
![image](https://github.com/getodk/collect/assets/3276264/bb71d544-7d59-408c-aa85-ea8f0cbe8160)

It's not a big deal we can file it as a separate issue.

dbemke commented:
@alyblenkin Could you have a look if the indentation in groups/repeats, questions and answers are ok? It seems as expected to me but it would be great if you could confirm and then we would also have a verified example that we could add to our tests :) 
![gruopsRepeats](https://github.com/getodk/collect/assets/56479916/ce0a0aff-2867-4c5c-abe3-ba3dea9b7ac3)

dbemke commented:
I'm not sure how a new line in the answer field should be displayed. This is an example of the hierarchy view with answers containing words with a new line between - the space between lines seems bigger than the space between the question and the answer.
![NewLineHierarchy](https://github.com/getodk/collect/assets/56479916/09e543d2-3a23-45ea-84be-2aa43ca065e1)

alyblenkin commented:
> @alyblenkin Could you have a look if the indentation in groups/repeats, questions and answers are ok? It seems as expected to me but it would be great if you could confirm and then we would also have a verified example that we could add to our tests :) 

@grzesiek2010 @seadowg feel free to jump in if I'm wrong, but this looks like the expected behaviour. 

> I'm not sure how a new line in the answer field should be displayed. This is an example of the hierarchy view with answers containing words with a new line between - the space between lines seems bigger than the space between the question and the answer.

@dbemke, you have such a great eye for this kind of detail 😍 I agree with you that the spacing should be the same. However, I don't think this is a show-stopper! If it is an easy fix, I think we should match the spacing between the question and answer (this might be a bit tight, but it's hard to say until we see it).

grzesiek2010 commented:
> you have such a great eye for this kind of detail 😍 I agree with you that the spacing should be the same. However, I don't think this is a show-stopper! If it is an easy fix, I think we should match the spacing between the question and answer (this might be a bit tight, but it's hard to say until we see it).

Please file separate issues for spacing and the margin between the group name and the list. Everything else seems to be fine.

dbemke commented:
**Tested with Success!**

Verified on devices with Android: 8.1, 10

**Verified cases:**
- hierarchy view of questions, answers, groups, repeats
- hierarchy view in light and dark mode
- landscape and portrait view
- scrolling the hierarchy view
- spacing, margins in the hierarchy view
- issues found in the PR will be filed separately

srujner commented:
**Tested with Success!**

Verified on devices with Android: 13
